exporters/secret/exporter.go

```diff
@@ -0,0 +1,156 @@
+package secret
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	"github.com/giantswarm/k8sclient/v5/pkg/k8srestconfig"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	fieldSelector = "type=kubernetes.io/tls"
+)
+
+type Config struct {
+	Namespaces []string
+}
+
+type Exporter struct {
+	cert      *prometheus.Desc
+	ctx       context.Context
+	k8sClient *kubernetes.Clientset
+	logger    micrologger.Logger
+
+	namespaces []string
+}
+
+func DefaultConfig() Config {
+	return Config{
+		Namespaces: []string{},
+	}
+}
+
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	e.logger.Log("info", "start collecting metrics")
+
+	// If no namespace whitelist is provided then we check all namespaces
+	if len(e.namespaces) == 0 {
+		allNamespaces, err := e.k8sClient.CoreV1().Namespaces().List(e.ctx, metav1.ListOptions{})
+		if err != nil {
+			e.logger.Log("error", microerror.Mask(err))
+		}
+
+		for _, ns := range allNamespaces.Items {
+			// We just need the namespace's name
+			nsName := ns.Name
+			e.namespaces = append(e.namespaces, nsName)
+		}
+	}
+
+	listOpts := metav1.ListOptions{
+		FieldSelector: fieldSelector,
+	}
+
+	// Range over namespaces
+	for _, namespace := range e.namespaces {
+		// Get secrets in namespace
+		secrets, err := e.k8sClient.CoreV1().Secrets(namespace).List(e.ctx, listOpts)
```

Comment:
did you think about using k8s ControllerRuntime client instead? we tend to use that for new code

Reply:
I wasn't aware of this as an option - I'll take a look into it and refactor.

Reply:
Can you point me in the direction of some example of this in action? Preferably something which isn't too hard for me to get my head around if possible!

Reply:
there is a Get request
https://github.com/giantswarm/awscnfm/blob/master/cmd/action/verify/netpol/curlrequest/runner.go#L81-L116

there is a create request
https://github.com/giantswarm/awscnfm/blob/master/cmd/action/create/netpol/defaultnetpol/runner.go#L111-L120

there is a List request:
https://github.com/giantswarm/awscnfm/blob/master/cmd/action/verify/kiam/podandsecret/runner.go#L123-L142

I, unfortunately, don't have any good example for creating the client
https://github.com/giantswarm/awscnfm/blob/master/pkg/client/tenant_cluster.go#L116-L127
it's hidden somewhere there

Reply:
I dont want to block you with that, but thought it would be cool to have it so proceed as you want